LPS-85785 V1 - Change DDMFormFieldValidation::_errorMessage and it's interface from String to LocalizedValue

IstvanD opened the pull request:
Hey @natocesarrego 

As asked on the PTR, could you please have a look at the changes so far?

Thank you and best regards,
István

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

natocesarrego commented on modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorHelper.java:
```diff
@@ -473,7 +475,16 @@ protected void validateField(
 
 		if (!valid) {
-			String errorMessage = ddmFormFieldValidation.getErrorMessage();
+			LocalizedValue errorMessageLocalizedValue =
```
Sometimes you name the variables as `errorMessage` and sometimes not, but I understand that, to keep consistency, you should name the variables always as `errorMessage`. After all, the type of the `errorMessage`, `LocalizedValue`, is already explicit in the class `DDMFormFieldValidation`.

IstvanD replied:
Got it, I changed this part. Now the `LocalizedValue` typed objects are the ones called `errorMessage`.

natocesarrego commented on modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorHelper.java:
```diff
@@ -473,7 +475,16 @@ protected void validateField(
+			LocalizedValue errorMessageLocalizedValue =
+				ddmFormFieldValidation.getErrorMessage();
+
+			String errorMessage = StringPool.BLANK;
```
This line makes the code of line 489 unreachable.

IstvanD replied:
Sure, I fixed this part.

natocesarrego commented on modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java:
```diff
@@ -559,16 +559,23 @@ protected void setDDMFormFieldTemplateContextValidation(
+		LocalizedValue errorMessageLocalizedValue =
+			ddmFormFieldValidation.getErrorMessage();
+
+		String errorMessage = null;
```
You should use the method `GetterUtil.getString`, as in the removed code, to set the `errorMessage` value and avoid a potential NPE.

IstvanD commented on modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/Sidebar/Sidebar.es.js:
```diff
@@ -243,24 +243,15 @@ class Sidebar extends Component {
-				type === 'validation'
-			)
-		);
+		return false;
```
Hi @natocesarrego 

I added this as a temporary workaround to prevent the error message field to be disabled in translations. Making the translated fields read-only was made in [LPS-93968](https://issues.liferay.com/browse/LPS-93968), and it prevents the error message to be localized.
My workaround is admittedly really simple and stupid, but I could not think of another way at the moment to allow me to make the fields localizable.

natocesarrego replied:
Thanks for your clarification, @IstvanD! 😄

IstvanD commented on modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/java/com/liferay/dynamic/data/mapping/form/builder/internal/context/DDMFormBuilderContextFactoryHelper.java:
```diff
@@ -290,12 +290,16 @@ protected Value doCreateDDMFormFieldValue(
+
 		jsonObject.put(
-			"errorMessage", ddmFormFieldValidation.getErrorMessage()
+			"errorMessage", localizedValueToJSONObject(
```
I removed the `localizedValueToJSONObject()` method since the logic is compact enough for it to be included in `doCreateDDMFormFieldValue()`.
`createOptions()` is a bit a bit different, as the options are stored as localized arrays:
`{"options" : {"en_US" : [{"label" : "value"}, {"label" : "value"} ...], "de_DE" : [{"label" : "value"}, {"label" : "value"} ...]}}
`
while my goal for errorMessage is to be structured like this:

`{ "errorMessage" : {"en_US" : "englisherror", "de_DE" : "germanerror", ...}}`

IstvanD commented:
Hey @natocesarrego 

I addressed your suggestions in the submitted code.
The validation error messages are now stored in the database as localized values. It can be verified by checking the `ddmstructure` table's `definition` column for newly added DDM Forms.


```
			"validation" : {
				"errorMessage" : {
					"en_US" : "oijoi"
				},
				"expression" : "isEmailAddress(TextField)"
			},
```


It only has entry for the default locale (en_US), because `DDMFormContextToDDMForm.getDDMFormFieldValidation() ` receives only a single string in `serializedValue`.

```
{"expression":"isEmailAddress(TextField)","enableValidation":true,"errorMessage":"invalid"}
```

The UI gas the get ready to accept localized values for the errorMessage field, but I'm not that familiar with handling localized inputs in the UI.

Do you now an good sample for this in the Forms UI?

Could you please also share your opinion on my changes so far?

Thank you and best regards,
István

IstvanD commented:
Hey @natocesarrego 

I reviewed the comments you made and made my changes accordingly. I pushed those changes to this pull request.

In regards to the front-end, I haven't yet started to implement that part, although that is the next step I'll  take. Currently I'm struggling to find a method to make the UI receive localized versions of the errorMessage in the Form Builder Sidebar. Because I'm not that familiar with the UI technology used in Forms, I don't know which files should I modify. I suppose those files a Javascript source files and Soy templates.
Are there any patterns for localized field intake in the current code that I could use and adapt?

As for `ValidationDDMFormFieldTemplateContextContributor`, I made the changes in advance with the localized errorMessages in mind but I may adjust it once the problem of the UI input and storage the localized errorMessage is solved.

Thank you and best regards,
István

natocesarrego commented:
Good job, @IstvanD!

In regards to the front-end code you should modify to complement your back-end solution, you should look [here](https://github.com/liferay/liferay-portal/tree/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/validation). On this folder, you'll find the Soy template and the Javascript files you're looking for. 😉 

Best.

IstvanD commented:
Hi @natocesarrego 

Thank you very much for your help so far!
I have tested the Form rendering (through the Form portlet) and it shows the validation error messages in their proper locales. Provided if they are store as proper localized values
`{ "errorMessage" : {"en_US" : "englisherror", "de_DE" : "germanerror", ...}}`

So it looks like the code [here](https://github.com/liferay/liferay-portal/tree/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/validation) is fine as it is.
However, I'm still struggling to save the translations for each available locale in the Forms builder.
When I'm changing the translation in the Forms builder, the value of the validation error message will stay the same as it was entered last (before switching languages) and won't change to the value belonging to the language that was switched to.
Also, when I'm saving the form, only the value for the default locale (in my case en_US) will be saved

`{ "errorMessage" : {"en_US" : "englisherror"}}`

Do you have an idea what I might be missing?

Thank you and best regards,
István

IstvanD commented:
Hi @natocesarrego 

Thank you very much for your help so far!
I have tested the Form rendering (through the Form portlet) and it shows the validation error messages in their proper locales. Provided if they are store as proper localized values
`{ "errorMessage" : {"en_US" : "englisherror", "de_DE" : "germanerror", ...}}`

So it looks like the code [here](https://github.com/liferay/liferay-portal/tree/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/validation) is fine as it is.
However, I'm still struggling to save the translations for each available locale in the Forms builder.
When I'm changing the translation in the Forms builder, the value of the validation error message will stay the same as it was entered last (before switching languages) and won't change to the value belonging to the language that was switched to.
Also, when I'm saving the form, only the value for the default locale (in my case en_US) will be saved

`{ "errorMessage" : {"en_US" : "englisherror"}}`

Do you have an idea what I might be missing?

Thank you and best regards,
István

natocesarrego commented:
Hey @IstvanD,

The Form Builder was migrated to Metal.js, hence you should look for the validation code implemented on Metal.js side. Please check the [Validation.es.js](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/metal/Validation/Validation.es.js) file, specifically the method [prepareStateForRender](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/metal/Validation/Validation.es.js#L15-L32). This method is executed once the Form Builder language is changed and also provides the new editing language (state.editingLanguageId).

Best.

IstvanD commented:
Hi @natocesarrego 

I [started to adapt](https://github.com/natocesarrego/liferay-portal/pull/147/commits/f2e151935d51a32e00e8544c0b276b4ac6100d25) Validation.es.js and Validation.soy to handle localized values for errorMessage, following Options.es.js and Options.soy as an example, as Options also deals with localization.

Options.es.js makes a lot of use of editingLanguageId and defaultLanguageId, utilizing it in functions such as getCurrentLocaleValue, getItems, syncEditingLanguageId etc and also in prepareStateForRender.

What I have done so far is to declare defaultLanguageId and editingLanguageId as parameters in the Vlaidation.soy template and turning errorMessage to an array in Validation.es.js, similar to items in Options.es.js.
However, that is as far as I got because I have no experience in development for metal.js.
Do you or someone else who is savvy in metal have suggestions for the next steps in the implementation?

Thank you in advance!

Best regards,
István

natocesarrego commented:
Hi @IstvanD,

Unfortunately, I don't have experience with Metal.js either, thus I'll need to ask Forms team for help. Maybe, would be better send you current implementation to a Forms front-end engineer implement the front-end side, but I need to discuss this alternative with the team first.

As soon as I get news, I let you know.

Best.

natocesarrego commented:
Hi @IstvanD,

I've just talked to @brunobasto and he's said to me that he'll take a look on this pull to assist you with the front-end part of this fix.

Best.

alinedoleron commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

alinedoleron commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

IstvanD commented:
Hi @natocesarregob, @brunobasto

Thank you very much!

natocesarrego commented:
Hi @IstvanD,

Just to let you know, @alinedoleron, another front-end engineer of Forms team, has taken this pull request to try to implement the front-end side of your solution, because she was working in a very similar issue. Therefore, from now on, she can provide the updates regarding her progress. In light of this, I think we can close this PR for now and reassign the LPS ticket to @alinedoleron, what do you think about it? Are you comfortable with this alternative?

Best.

IstvanD commented:
Hi @natocesarrego , @alinedoleron 

I see. Many thanks for helping me out with this. Please let me know if you have any questions about the solution so far.

Best regards,
István

natocesarrego commented:
@IstvanD, I'm closing this pull request, once this issue isn't on my side anymore.

Best.
